bcfs/src/bcfs.rs

```diff
@@ -11,113 +11,100 @@ use wasi_types::{
+    Result,
 };
 
 pub struct BCFS<A: Address, M: AccountMeta> {
-    files: Vec<Option<Filelike<A>>>,
-    blockchain_name: String,
-    home_dir: PathBuf,
+    files: Vec<Option<File<A>>>,
+    home_addr: A,
     _account_meta: std::marker::PhantomData<M>,
 }
 
 impl<A: Address, M: AccountMeta> BCFS<A, M> {
     /// Creates a new ptx FS with a backing `ptx` and hex stringified
     /// owner address.
     pub fn new<S: AsRef<str>>(
-        ptx: &mut dyn PendingTransaction<Address = A, AccountMeta = M>,
+        home_addr: A,
+        // ptx: &mut dyn PendingTransaction<Address = A, AccountMeta = M>,
         blockchain_name: S,
     ) -> Self {
-        let mut home_dir = PathBuf::from("/opt");
-        home_dir.push(blockchain_name.as_ref());
-        home_dir.push(ptx.address().path_repr());
-
         Self {
-            files: vec![
-                Some(Filelike::File(File::stdin())),
-                Some(Filelike::File(File::stdout())),
-                Some(Filelike::File(File::stderr())),
-                Some(Filelike::File(File::log())),
-            ],
-            blockchain_name: blockchain_name.as_ref().to_string(),
-            home_dir,
+            files: File::defaults(blockchain_name.as_ref()),
+            home_addr,
             _account_meta: std::marker::PhantomData,
         }
     }
 
+    pub fn prestat(
+        &mut self,
+        _ptx: &mut dyn PendingTransaction<Address = A, AccountMeta = M>,
+        fd: Fd,
+    ) -> Result<&Path> {
+        match &self.file(fd)?.kind {
+            FileKind::Directory { path } => Ok(path),
+            _ => Err(ErrNo::BadF),
+        }
+    }
+
     pub fn open(
         &mut self,
         ptx: &mut dyn PendingTransaction<Address = A, AccountMeta = M>,
-        curdir: Option<Fd>,
+        curdir: Fd,
         path: &Path,
         open_flags: OpenFlags,
         fd_flags: FdFlags,
     ) -> Result<Fd> {
-        if open_flags.contains(OpenFlags::DIRECTORY) | curdir.is_some() {
-            // The virutal filesystem does not yet support directories.
+        if open_flags.contains(OpenFlags::DIRECTORY) {
+            // The virutal filesystem does not yet allow opening directories.
             return Err(ErrNo::NotSup);
         }
 
-        let path = self.canonicalize_path(path)?;
+        match &self.file(curdir)?.kind {
+            FileKind::Directory { .. } => (),
+            _ => return Err(ErrNo::BadF),
+        };
 
-        if path == Path::new("/log") {
-            if open_flags.intersects(OpenFlags::CREATE | OpenFlags::EXCL) {
-                return Err(ErrNo::Exist);
-            }
-            if !fd_flags.contains(FdFlags::APPEND)
-                || open_flags.intersects(OpenFlags::TRUNC | OpenFlags::DIRECTORY)
-            {
-                return Err(ErrNo::Inval);
+        let mut file_exists = true;
+        let file_kind = match self.canonicalize_path(curdir, path)? {
+            (None, path) if path == Path::new("log") => FileKind::Log,
```

Comment:
should these get promoted to constants?

Reply:
maybe, but it's only used here. tbh the actual constant is the entire path resolution algorithm